### PR TITLE
feat: add metrics-server addon

### PR DESCRIPTION
**Pull request by jenting**
## Why is this PR needed?

Fixes https://github.com/SUSE/avant-garde/issues/1336

## What does this PR do?

Add metrics-server addon:
- Generate metrics-server certificate signed by Kubernetes CA
- set deployment replicas=2 for HA
- set `revisionHistoryLimit=3` in deployment manifest #981
- set label `caasp.suse.com/skuba-replica-ha: "true"` in deployment manifest #951 

## Anything else a reviewer needs to know?

`kubectl top --sort-by` [issue](https://github.com/kubernetes/kubernetes/issues/81270) and _might_ release in `1.18` version.

## Info for QA

This is info for QA so that they can validate this. This is **mandatory** if this PR fixes a bug.
If this is a new feature, a good description in "What does this PR do" may be enough.

### Related info

Info that can be relevant for QA:
* link to other PRs that should be merged together
* link to packages that should be released together
* upstream issues

### Status **BEFORE** applying the patch

N/A

### Status **AFTER** applying the patch

- greenfield
```
skuba cluster init --control-plane <IP/FQDN> <cluster-name>
skuba node bootstrap -s -u <user> -t <IP/FQDN> master0
skuba node join --role worker -s -u <user> -t <IP/FQDN> worker0

kubectl top nodes
kubectl top pods
```

- brownfield
```
skuba addon upgrade plan
skuba addon upgrade apply

kubectl top nodes
kubectl top pods
```

## Docs

If docs need to be updated, please add a link to a PR to https://github.com/SUSE/doc-caasp.
At the time of creating the issue, this PR can be work in progress (set its title to [WIP]),
but the documentation needs to be finalized before the PR can be merged. 

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

Signed-off-by: JenTing Hsiao <jenting.hsiao@suse.com>

### CODE REVIEW

File: internal/pkg/skuba/addons/metricsserver.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	reader := bufio.NewReader(f)
+	content, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return ""
+	}
+
+	return base64.StdEncoding.EncodeToString(content)
+}
```
Comment (jenting) [inline internal/pkg/skuba/addons/metricsserver.go]: The `caBundle` renders the Kubernetes CA with base64 encoded.

When `skuba cluster init`, at this time Kubernetes CA has not generated yet, render the `caBundle` empty string.

When `skuba bootstrap` and `skuba addon upgrade apply`, the Kubernetes CA file had generated. When apply addon, the manifest would be re-render again. So the `caBundle` rendered with base64 Kubernetes CA.

**Comment by jenting** [comment]
Ready to review.

File: internal/pkg/skuba/metricsserver/metricsserver.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+const (
+	certCommonName = "metrics-server.kube-system.svc" // DO NOT CHANGE THE CN
+	secretCertName = "metrics-server-cert"
+)
+
+// CreateCert creates a signed certificate for metrics-server
+// with kubernetes CA certificate and key
+func CreateCert(client clientset.Interface, pkiPath string) error {
```
Comment (innobead) [inline internal/pkg/skuba/metricsserver/metricsserver.go]: These two functions seem able to reuse in somewhere except constants. What do u think to move to a common place?

File: internal/pkg/skuba/kubernetes/versions.go [hunk 3]
```diff
@@ -91,11 +92,12 @@ var (
...
-				Cilium:  &AddonVersion{"1.5.3", 1},
-				Kured:   &AddonVersion{"1.3.0", 2},
-				Dex:     &AddonVersion{"2.16.0", 5},
-				Gangway: &AddonVersion{"3.1.0-rev4", 4},
-				PSP:     &AddonVersion{"", 1},
+				Cilium:        &AddonVersion{"1.5.3", 1},
+				Kured:         &AddonVersion{"1.3.0", 2},
+				Dex:           &AddonVersion{"2.16.0", 5},
+				Gangway:       &AddonVersion{"3.1.0-rev4", 4},
+				MetricsServer: &AddonVersion{"0.3.6", 0},
```
[the remark this answers is not in the document]
Comment (jenting) [reply internal/pkg/skuba/kubernetes/versions.go]: I had the same question as you. It starts at 0.

https://github.com/SUSE/skuba/blob/release-caasp-4.0.2/internal/pkg/skuba/kubernetes/versions.go#L95

File: internal/pkg/skuba/metricsserver/metricsserver.go (hunk 2 above)
Comment (jenting) [reply internal/pkg/skuba/metricsserver/metricsserver.go]: I'd like to refactor it when doing certificate management feature. WDYT?

Comment (innobead) [reply internal/pkg/skuba/metricsserver/metricsserver.go]: Ha sure ;)

File: internal/pkg/skuba/addons/metricsserver.go (hunk 1 above)
Comment (c3y1huang) [reply internal/pkg/skuba/addons/metricsserver.go]: @jenting this information is useful. Suggest to add as function comment.

Comment (c3y1huang) [reply internal/pkg/skuba/addons/metricsserver.go]: @jenting this information is useful. Suggest to add as function comment.

**Comment by jenting** [comment]
rebase to master branch and force push, CI green :heavy_check_mark:

File: internal/pkg/skuba/addons/metricsserver.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+func GetMetricsServerImage(imageTag string) string {
+	return images.GetGenericImage(skubaconstants.ImageRepository, "metrics-server", imageTag)
+}
+
+func (renderContext renderContext) MetricsServerImage() string {
+	return GetMetricsServerImage(kubernetes.AddonVersionForClusterVersion(kubernetes.MetricsServer, renderContext.config.ClusterVersion).Version)
+}
+
+// CABundle returns base64 encoded Kubernetes CA certificate
+//
```
Comment (c3y1huang) [inline internal/pkg/skuba/addons/metricsserver.go]: @jenting ,the comment block looks a little detached. Could you remove the blank lines?

Comment (jenting) [reply internal/pkg/skuba/addons/metricsserver.go]: sure~

**Comment by kravciak** [comment]
Our [documentation](https://susedoc.github.io/doc-caasp/master/single-html/caasp-architecture/#_testing_poc) says minimal configuration is 1 master 1 worker.

Guessing from `metrics` in PR title it's this one that breaks our tests (we check that all pods are running after cluster creation).

```
skuba:master λ kubectl get pods -n kube-system -l app=metrics-server
NAME                              READY   STATUS    RESTARTS   AGE
metrics-server-554b946d44-7lwg7   1/1     Running   0          16m
metrics-server-848d94b477-8blgc   0/1     Pending   0          14m
metrics-server-848d94b477-jxnbq   1/1     Running   0          14m

skuba:master λ kubectl describe pod -n kube-system metrics-server-848d94b477-8blgc | tail -10
Tolerations:     node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod anti-affinity rules, 1 node(s) didn't satisfy existing pods anti-affinity rules, 2 node(s) didn't match pod affinity/anti-affinity.
```

**Comment by jenting** [comment]
I this it's related to #951 cc @c3y1huang

**Comment by jenting** [comment]
> Our [documentation](https://susedoc.github.io/doc-caasp/master/single-html/caasp-architecture/#_testing_poc) says minimal configuration is 1 master 1 worker.
> 
> Guessing from `metrics` in PR title it's this one that breaks our tests (we check that all pods are running after cluster creation).
> 
> ```
> skuba:master λ kubectl get pods -n kube-system -l app=metrics-server
> NAME                              READY   STATUS    RESTARTS   AGE
> metrics-server-554b946d44-7lwg7   1/1     Running   0          16m
> metrics-server-848d94b477-8blgc   0/1     Pending   0          14m
> metrics-server-848d94b477-jxnbq   1/1     Running   0          14m
> 
> skuba:master λ kubectl describe pod -n kube-system metrics-server-848d94b477-8blgc | tail -10
> Tolerations:     node-role.kubernetes.io/master:NoSchedule
>                  node.kubernetes.io/not-ready:NoExecute for 300s
>                  node.kubernetes.io/unreachable:NoExecute for 300s
>   Type     Reason            Age        From               Message
>   ----     ------            ----       ----               -------
>   Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
>   Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
>   Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't match pod anti-affinity rules, 1 node(s) had taints that the pod didn't tolerate.
>   Warning  FailedScheduling  <unknown>  default-scheduler  0/2 nodes are available: 1 node(s) didn't match pod anti-affinity rules, 1 node(s) didn't satisfy existing pods anti-affinity rules, 2 node(s) didn't match pod affinity/anti-affinity.
> ```

I this it's related to #951 cc @c3y1huang

**Comment by innobead** [comment]
@kravciak could you help guide us where to see the complete logs (I assume you run by validator from jenkins?)? This will save us time to check the root cause soon. Thanks.

Interesting part is the replica is 2 but here the output is 3, 1 is pending w/ some untolerated taint.

**Comment by c3y1huang** [comment]
The issue is coming from setting replica size to 2 causing scheduler trying to schedule pod when node (worker) wasn't ready: 
1 node(s) didn't match pod anti-affinity rules, `1 node(s) had taints that the pod didn't tolerate`.

With replica set to 2, affinity rule would have already set all node. 1 Pod has been scheduled on master node, so we see the complains about it didn't match anti-affinity rule. But the scheduler also couldn't place the pod on worker node because is not ready (tainted). This leaving the other pod `Pend` to wait until taint is removed from worker node:
```
Tolerations:     node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
```
When the worker node is ready, then taint got removed and scheduler is able to place pod onto worker node (2 pods running).  And somehow the pod that in `Pending` was not removed :thinking: 

For quick fix for metrics-server I would suggest to set replica size to 3 in manifest or remove the `skuba-replica-ha`.

**Pull request merged**